### PR TITLE
Make sure MKL uses GNU OpenMP.

### PR DESCRIPTION
**Pull request by abergeron**

### CODE REVIEW

**Comment by abergeron** [comment]
So this doesn't work for MKL 2017 so we need an alternate solution that handles this.

Unfortunately all means of checking the current version of MKL already load the library and prevent this fix from working.

Also I just realized that this is pointless if the user has already loaded numpy before theano since MKL is already loaded so it's too late.

**Comment by abergeron** [comment]
I've amended the commit to be more extensive.  Also instead of attempting to fix the problem, we just raise an error if it is detected.

There is an undocumented flag to disable the check if required.

**Comment by abergeron** [comment]
Just to check if the mac and windows checkers need special config.

jenkins test mac windows

**Comment by abergeron** [comment]
jenkins test mac windows

File: theano/configdefaults.py [hunk 1]
```diff
@@ -1238,6 +1238,29 @@ def is_valid_check_preallocated_output_param(param):
...
+    import os
+    if ('MKL_THREADING_LAYER' in os.environ and
+            os.environ['MKL_THREADING_LAYER'] == 'GNU'):
+        return
+    try:
+        import mkl
+        if '2018' in mkl.get_version_string():
+            raise RuntimeError('To use MKL 2018 with Theano you MUST set "MKL_THREADING_LAYER=GNU" in your environement.')
+    except ImportError:
+        raise RuntimeError("""
```
Comment (nouiz) [inline theano/configdefaults.py]: I just want to clarity, with this PR, now the mkl package is mandatory with conda. If the user don't use conda, and the string "mkl" isn't in the flag, then he don't need to change his environment. Can you confirm that?

Comment (abergeron) [reply theano/configdefaults.py]: If the string mkl isn't in the flag, then nothing changes.

However if you use mkl and are not in conda then you either need to obtain the mkl-service package somehow, set MKL_THREADING_LAYER=GNU or disable the check with blas.check_openmp=False

**Pull request merged**